### DISCUSSION TITLE
Perspective Distortion on a Polygon

### DISCUSSION THREAD

**Discussion by fcuenca4** [ImageMagick/ImageMagick#6470] [Help]
### ImageMagick version

ImageMagick 7.1.1-12

### Operating system, version and so on

MAC

### Description

Hello, I'm currently looking for assistance in implementing a perspective transformation on a given image.

Given an image, my task is to carry out a perspective distortion that maps to a polygon with the following coordinates:
```		
TopLeft:     {X: 53.7871094, Y: 676.910156},
TopRight:    {X: 998.980384, Y: 132.974609},
BottomRight: {X: 1992.25, Y: 707.041016},
BottomLeft:  {X: 1047.50977, Y: 1250.70117},
Height:      1280,
Width:       2106,
```
![input](https://github.com/ImageMagick/ImageMagick/assets/9162072/29c85fcd-b82f-4f56-85e6-2a63cf559e0e)

I have attempted to apply the distortion by utilizing the following command:

```
convert input.png -alpha set -resize 2106x1280 -matte -virtual-pixel transparent -distort Perspective '0 0 53.7871094 676.910156 960 0 484.64143726893224 428.9640580902445 960 960 1229.8703566623906 859.4545027542885 0 960 799.1709321927204 1107.3064385968255' output.jpg
```
This command results in a perspective transformation that looks like this:
![output](https://github.com/ImageMagick/ImageMagick/assets/9162072/2014ff21-f800-48e0-bd6d-cb8d13fc50d9)

However, the output is not as expected. The desired outcome is something more akin to this:
<img width="453" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/9162072/f14c2b22-eb14-4ffd-8888-7a5dfcefa7ca">

I would appreciate any guidance or suggestions on how I could adjust my approach to better achieve my desired outcome. Thank you in advance for your help!

**Comment by snibgo** [comment]
I suggest using `magick` instead of `convert`. This will give you the IM v7 syntax instead of the inferior v6 syntax.

JPG can't record transparency. If you want a transparent background, you should save to a format that can record transparency, such as TIFF or PNG.

`-matte` is deprecated. Use `-alpha set` instead. As you already have that, you don't need to repeat it.

You resize the input to `2106x1280` without changing the aspect ratio. The image is square, so it is resized to 1280x1280. If you want to make the output exactly 2106x1280, use an exclamation mark: `2106x1280!`. (This may need quoting or escaping in your command.)

Does that fix all the problems?

**Reply by fcuenca4** [reply]
not really - using +distort Perspective did the trick. But now I'm having issues applying this mask: 
![PBK-1_MASK](https://github.com/ImageMagick/ImageMagick/assets/9162072/03170d17-a41e-4d41-bd25-cc9249185a1f)

on the following proyected image: 
![skewed](https://github.com/ImageMagick/ImageMagick/assets/9162072/1ea9bbc6-ea1a-49bc-ad20-24d702d86906)

`convert -compose over  PBK-1_MASK.png skewed.png -composite masked.png` generates the following output: 
![withMask](https://github.com/ImageMagick/ImageMagick/assets/9162072/9f7d6ac1-183c-42c1-9981-bc2a00e51bae)

I'm not sure what am I missing 😕

**Comment by fmw42** [comment]
Your mask should be binary white and black, not transparent.  On IM 7, I suggest you use magic not convert. Read the images first, then the mask, then your compose settings, then the composite operator.

`magick skew.png  PBK-1_MASK.png -compose over -composite masked.png`

This assumes your mask and image are the same size and aligned with each other.

**Comment by snibgo** [comment]
> But now I'm having issues applying this mask: ... on the following proyected image: ... I'm not sure what am I missing 😕

You show a command and an output, but you don't say what result you want. What change should occur to your image?

**Reply by fcuenca4** [reply]
should look something like this (sorry, it's a different image but the idea is still the same)
<img width="660" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/9162072/a02c9aff-25c8-4983-8750-540d288249aa">

for some reason the images don't fit: 
Mask is `2106x1280` whilst the distorted is `1179x681` but even resizing the last one to the same size doesn't seem to be the right fit.

Here's my attempt to resize the image using `magick skewed.png -resize 2106x1280! skewed2.png`. 
Then, composing over generates the following image which is not correct
![masked](https://github.com/ImageMagick/ImageMagick/assets/9162072/f9fe41a1-29e5-47f3-a39f-f3dee9d3122f)

If I don't resize the image, this happens: https://user-images.githubusercontent.com/9162072/252400014-9f7d6ac1-183c-42c1-9981-bc2a00e51bae.png

**Comment by snibgo** [comment]
So, you want to copy the transparency ("opacity") from the mask to the image. Use "`-compose CopyOpacity -composite`" for that.

The image has an offset, so you need "+repage" to remove that. The mask is surrounded by transparent pixels. Us "-trim" to remove them.
```
magick fcuencaImg.png fcuencaMask.png -trim +repage -compose CopyOpacity -composite out.png
```
The result looks roughly correct, but not exactly. You should adjust the mask so the boundary is entirely inside the opaque image.

**Reply by fcuenca4** [reply]
And that solves it, many thanks for your help! :)

**Reply by fcuenca4** [reply]
> So, you want to copy the transparency ("opacity") from the mask to the image. Use "`-compose CopyOpacity -composite`" for that.
> 
> 
> 
> The image has an offset, so you need "+repage" to remove that. The mask is surrounded by transparent pixels. Us "-trim" to remove them.
> 
> ```
> 
> magick fcuencaImg.png fcuencaMask.png -trim +repage -compose CopyOpacity -composite out.png
> 
> ```
> 
> The result looks roughly correct, but not exactly. You should adjust the mask so the boundary is entirely inside the opaque image.
> 
> 

@snibgo 
If you allow me a last question, in your expertise, what's the most efficient way to apply these commands? Is there any way to avoid many writes to disks?
Thanks!

**Reply by fcuenca4** [reply]
@snibgo I'm still having some difficulties in composing the subsequent lighting layer, and I suspect it might be due to the elimination of the offsets. All of the static layers: tint, mask, lighting, etc. maintain fixed dimensions of 2106x1280.

Could this be connected to the trim/repage operations?
![PBK-1_LIGHTING](https://github.com/ImageMagick/ImageMagick/assets/9162072/022dfa84-5c63-4456-b507-74c4b6fac13b)

As you pointed out in your prior comment, without applying -trim and +repage the alignment isn't perfect. But with the -gravity Center operation, it seems to align properly.

Without -gravity Center:
![withMask](https://github.com/ImageMagick/ImageMagick/assets/9162072/2f00bf90-c627-4069-8ffe-15d63221aca1)

with -gravity Center
![withMask](https://github.com/ImageMagick/ImageMagick/assets/9162072/e429024b-f192-4aaf-9c3c-fa1903dafb6e)

I'm trying to get a result as it follows
![finalImageRender](https://github.com/ImageMagick/ImageMagick/assets/9162072/8fa68c7b-ae1d-43f5-b039-b6bcf7e6cd81)

Apologies for adding another query to this thread, but I must admit image processing isn't my strongest area.
Thanks,
Fran

**Reply by fcuenca4** [reply]
`magick out.png PBK-1_LIGHTING.png -compose over +repage -trim -gravity north  -composite test_lighting.png` does improve things, but I'm trying to find a way to make this process generic; as my layers and projections will change depending on the product I'm selecting.  They all have the same properties 
![test_lighting](https://github.com/ImageMagick/ImageMagick/assets/9162072/e4ca71c4-c99d-43a7-9a0e-b788a1b9cee7)

**Comment by snibgo** [comment]
Reading and writing PNG is slower than some other formats. PNG compression is designed for graphics, rather than photos. In my experience, TIFF is faster, even though the TIFF file may be larger. Performance tradeoffs will depend on CPU and disk performance. These days, I use SSD so reading/writing large files is limited by CPU performance for complicated formats such as PNG, rather than by disk performance.

MIFF and MPC can be faster than TIFF. These formats are propriety to ImageMagick, so are useful within a chain of image processing rather than as end-products. They are especially useful if, for example, you use the same mask many times.

When you have a chain of IM commands that use temporary files to send the output of one command to the input of another, it is more efficient to combine the commands so no temporary storage is needed.

**Comment by snibgo** [comment]
It is possible to distort a rectangular image so the edges follow the curved edges of the mask image. That is a bit complex, so I'll ignore that.

> All of the static layers: tint, mask, lighting, etc. maintain fixed dimensions of 2106x1280.

That is useful. You have shown three image so far:
```
magick fcuenc* info:

fcuencaImg.png[0] PNG 1179x681 1179x681+87+268 8-bit TrueColorAlpha sRGB 880614B 0.000u 0:00.001
fcuencaLighting.png[1] PNG 2106x1280 2106x1280+0+0 8-bit PaletteAlpha sRGB 256c 176624B 0.000u 0:00.003
fcuencaMask.png[2] PNG 2106x1280 2106x1280+0+0 8-bit PaletteAlpha sRGB 256c 11432B 0.000u 0:00.005
```
`Lighting` and `Mask` are the same size. `Img` is smaller, and with an offset. `-trim +repage` helps, but is too crude for accurate alignment. It is better to extend `Img` by the appropriate amount to make it align with the others. This turns out to be "`-extent 2106x1280-84-269`". The dimensions `2106x1280` come from the size of the other images. I found the offsets `-84-269` by manual trial and error.

But you wouldn't want to do this for every input image. A better solution is to do the distort perspective so the corners of the input image are placed appropriately. Taking `PreDistort` as the one you showed in your first post (Windows BAT syntax):
```
magick ^
  fcuencaPreDistort.png ^
  -set option:Wm1 %%[fx:w-1] ^
  -set option:Hm1 %%[fx:h-1] ^
  -background None -extent 2106x1280+0+0 ^
  -virtual-pixel None ^
  -distort perspective "0,0,81,656 %%[Wm1],0 750,270 0,%%[Hm1],590,955 %%[Wm1],%%[Hm1],1260,558" ^
  +repage ^
  d.png
```
The command records width and height minus one in Wm1 and Hm1, then extends the image to 2106x1280. Then it distorts the extended image to put the original corners at (81,656) etc. I tweaked the output coordinates so the distorted image is entirely outside the mask. We can check the result:
```
magick ^
  d.png ^
  ( fcuencaMask.png -channel A -evaluate Multiply 0.5 ) ^
  -compose Over -composite ^
  check.png
```
`check.png` shows that the distorted image is entirely outside the mask:
![fi](https://github.com/ImageMagick/ImageMagick/assets/33812368/a3f8f89f-e0df-46e8-8d0e-b54ced3d3f09)

Now we can combine `d.png`, and `Mask` and `Lighting`:
```
magick ^
  d.png ^
  fcuencaMask.png^
  -compose CopyOpacity -composite ^
  fcuencaLighting.png^
  -compose Over -composite ^
  out.png
```
`out.png` is:
![out](https://github.com/ImageMagick/ImageMagick/assets/33812368/e2e4cf65-aefb-4926-bda2-7f230af34abf)

We can combine the commands:
```
magick ^
  fcuencaPreDistort.png ^
  -set option:Wm1 %%[fx:w-1] ^
  -set option:Hm1 %%[fx:h-1] ^
  -background None -extent 2106x1280+0+0 ^
  -virtual-pixel None ^
  -distort perspective "0,0,81,656 %%[Wm1],0 750,270 0,%%[Hm1],590,955 %%[Wm1],%%[Hm1],1260,558" ^
  +repage ^
  fcuencaMask.png^
  -compose CopyOpacity -composite ^
  fcuencaLighting.png^
  -compose Over -composite ^
  out2.png
```

This will work for any input image. However, portrait format inputs will be distorted, which might look bad. Such inputs can either be cropped to landscape format, or extended, before continuing as above.

**Reply by fmw42** [reply]
Hoe do you get "original corners at (81,656)"?  Is that not manually selected? If not, then how?

**Comment by snibgo** [comment]
I chose the output coordinates (81,656) etc manually. If a different mask is used,  different output coordinates would be needed.

The output coordinates could be calculated automatically from the mask. For example:

1. Find the four corners of the mask.
2. Do a perspective transformation to move the four corners to a rectangle. This makes the edges of the mask roughly horizontal and vertical.
3. Find the bounding box of the result of (2). This gives coordinates of four corners.
4. Apply the inverse of the perspective transformation from (2) to the numerical coordinates of the four corners. The numerical results are the required output coordinates.

That needs a bit of tricky mathematics. Instead, we can do it graphically:

1. Find the four corners of the mask.
2. Do a perspective transformation to move the four corners to a rectangle. This makes the edges of the mask roughly horizontal and vertical.
3. Find the bounding box of the result of (2), and draw that as a rectangle.
4. Apply the inverse of the perspective transformation from (2) to the rectangle drawn in (3).
5. Find the four corners of the result from (4). Those four sets of coordinates are the required output coordinates.

**Comment by dost21** [comment]
super
